### PR TITLE
ARROW-9429: [Python] ChunkedArray.to_numpy

### PR DESCRIPTION
**Pull request by kszucs**
While it technically is still using pandas during the conversion it exposes the to_numpy() method.

Also refactored the chunked array construction to support more flexible input, see the test case.

### CODE REVIEW

**Review by jorisvandenbossche — APPROVED** [review]
> While it technically is still using pandas during the conversion it exposes the to_numpy() method.

I think it is actually not using pandas (although the `ConvertChunkedArrayToPandas` name is certainly confusing in that case, but I thought this also worked without pandas being present. It just converts the arrow data to an array or dict that is in the appropriate form to convert to pandas)

**Comment by pitrou** [comment]
@kszucs Can you rebase this PR? Hopefully it will fix AppVeyor.

File: python/pyarrow/table.pxi [hunk 1]
```diff
@@ -226,24 +226,34 @@ cdef class ChunkedArray(_PandasConvertible):
...
 
         if self.type.id == _Type_EXTENSION:
-            return (
-                chunked_array(
-                    [self.chunk(i).storage for i in range(self.num_chunks)]
-                ).__array__(dtype)
-            )
+            storage_array = chunked_array([
+                chunk.storage for chunk in self.iterchunks()
+            ])
```
Comment (pitrou) [inline python/pyarrow/table.pxi:245]: Does this work if there's zero chunk, or do you need to pass the type explicitly?

File: python/pyarrow/table.pxi [hunk 2]
```diff
@@ -425,31 +438,35 @@ def chunked_array(arrays, type=None):
...
+        if type is None:
+            # it allows more flexible chunked array construction from to coerce
+            # subsequent arrays to the firstly inferred array type
+            # it also spares the inference overhead after the first chunk
+            type = arr.type
         else:
-            arr = array(x, type=type)
+            if arr.type != type:
+                raise ArrowInvalid(
+                    "Each array chunks must have type {}".format(type)
```
Comment (pitrou) [inline python/pyarrow/table.pxi:458]: Either "All array chunks" (plural) or "Each array chunk" (singular).

File: python/pyarrow/table.pxi (hunk 1 above)
Comment (kszucs) [reply python/pyarrow/table.pxi:245]: Need to pass the type explicitly, fixing it.

**Comment by kszucs** [comment]
@pitrou rebased and addressed your comments

**Review by pitrou — APPROVED** [review]
+1, will merge

**Pull request closed without merging**